src/Microsoft.Azure.Batch.SoftwareEntitlement/NodeEntitlementReader.cs

```diff
@@ -0,0 +1,144 @@
+using System;
+using System.Globalization;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
+using Microsoft.Azure.Batch.SoftwareEntitlement.Common;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Azure.Batch.SoftwareEntitlement
+{
+    public class NodeEntitlementReader
+    {
+        private readonly string _expectedAudience;
+        private readonly string _expectedIssuer;
+        private readonly SecurityKey _signingKey;
+        private readonly SecurityKey _encryptionKey;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TokenReader"/> class
+        /// </summary>
+        /// <param name="signingKey">Optional key to use when verifying the signature of the token.</param> 
+        /// <param name="encryptingKey">Optional key to use when decrypting the token.</param> 
+        public NodeEntitlementReader(
+            string expectedAudience,
+            string expectedIssuer,
+            SecurityKey signingKey = null,
+            SecurityKey encryptingKey = null)
+        {
+            _expectedAudience = expectedAudience;
+            _expectedIssuer = expectedIssuer;
+            _signingKey = signingKey;
+            _encryptionKey = encryptingKey;
+        }
+
+        public Errorable<NodeEntitlements> ReadFromToken(string tokenString)
```

Comment:
This method feels a bit long to read comfortably - what can we do to make it shorter?

Perhaps extract `Errorable<ClaimsPrincipal> ReadPrincipal(string token)` from lines 38-72?